Add file history view

ddoemonn opened the pull request:
Closes #16827

Release Notes:

- Added: File history view accessible via right-click context menu on files in the editor or project panel. Shows commit history for the selected file with author, timestamp, and commit message. Clicking a commit opens a diff view filtered to show only changes for that specific file.

<img width="1293" height="834" alt="Screenshot 2025-11-11 at 16 31 32" src="https://github.com/user-attachments/assets/3780d21b-a719-40b3-955c-d928c45a47cc" />
<img width="1283" height="836" alt="Screenshot 2025-11-11 at 16 31 24" src="https://github.com/user-attachments/assets/1dc4e56b-b225-4ffa-a2af-c5dcfb2efaa0" />

ddoemonn committed:
feat(git-ui): add file history timeline view

ddoemonn committed:
fix: formatting and clippy errors

ddoemonn committed:
Merge branch 'zed-industries:main' into feat/file-history

0x2CA commented:
> Hi! Just passing by and wondering if this is lazy loading or not? If a file has a history that is too large (old repos like Linux or Firefox), it might slow down the UI.

I've taken a general look at the implementation, and it's simply using git asynchronously to fetch the history and then parsing the git output

need to change to stream loading, in batches, and display dynamically while the user scrolls

ddoemonn committed:
feat(git): add file history view with remote support

ddoemonn committed:
Merge branch 'feat/file-history' of github.com:ddoemonn/zed into feat/file-history

ddoemonn commented:
Hi @cameron1024 

I made some updates can you check it out?

Thanks!

cameron1024 reviewed (commented):
Backend stuff is looking better, just a few small issues 🎉

I think for now, we're just waiting on a design review - I'll let you know when I have more info. Apologies for the delay 😅

ddoemonn committed:
refactor: use hardcoded UUID, simplify file_history call, flatten nested if-lets

ddoemonn commented:
@cameron1024 thanks for feedbacks, I learned so much from this PR 😅

ddoemonn committed:
feat(file-history): re-design UI and logic

ddoemonn committed:
refactor: some UI adjustments on file history view

ddoemonn committed:
fix: formatting

ddoemonn committed:
Merge branch 'main' into feat/file-history

ddoemonn committed:
fix: redundant clone

ddoemonn committed:
Merge branch 'feat/file-history' of github.com:ddoemonn/zed into feat/file-history

ddoemonn committed:
fix: clippy errors

ddoemonn committed:
fix: proto

ddoemonn committed:
fix: formatting

ddoemonn committed:
fix: proto reserve

cameron1024 commented on crates/git_ui/src/commit_view.rs:
```diff
@@ -242,11 +275,398 @@ impl CommitView {
+                    .color(Color::Muted)
+                    .size(IconSize::Medium)
+                    .into_element()
+                    .into_any(),
```
Actually just noticed the `else` branch also has the same code. It's possible to extract that with a big `if let ...`, but it might be nicer to just pull out a `Self::fallback_commit_avatar()` function for this case

Commit:
refactor: commit delimiter with concat

Commit:
refactor: extract fallback avatar helper

Commit:
refactor: scope pagination vars to if-blocks

Commit:
test: add process_github_issues tests

cameron1024 commented on crates/git/src/repository.rs:
```diff
@@ -1421,6 +1444,94 @@ impl GitRepository for RealGitRepository {
+
+                // Add pagination arguments if needed
+                let skip_str;
+                let limit_str;
```
Ah, didn't spot that, in that case feel free to keep it how it is :)

ddoemonn commented:
@cameron1024 quick question should we remove this Metada field from the commit_view? 
<img width="786" height="551" alt="Screenshot 2025-11-28 at 17 29 29" src="https://github.com/user-attachments/assets/4539c9d3-debe-419d-b076-785faf89b7ad" />

Commit:
refactor: commented out metadata fields on commit view

Commit:
formatting

ddoemonn committed:
fix: comment

ddoemonn committed:
Merge branch 'main' of github.com:ddoemonn/zed into feat/file-history

ddoemonn committed:
fix: update RunGitHook field number in Envelope message

ddoemonn committed:
Merge branch 'main' into feat/file-history

ddoemonn committed:
refactor: Adjust `RepositoryState` destructuring, reassign proto message IDs, and optimize scroll handle usage.

ddoemonn committed:
fix: formatting

ddoemonn committed:
fix: reserve field 396 and reassign ExternalExtensionAgentsUpdated to 401

Field 396 was previously used by GitDeleteBranch and must be reserved to
maintain protobuf compatibility.

ddoemonn commented:
@cameron1024 the PR is ready to review!

cameron1024 committed:
chore: remove old commented-out sections

cameron1024 reviewed (approved):
Amazing! Thanks so much for working on this - I think we can merge it now! 🎉

BTW, I pushed a commit just cleaning up a few old sections of code that were commented out.  

Can't wait to start using this, thanks again 🙏

cameron1024 committed:
Merge branch 'main' into feat/file-history

cameron1024 committed:
fix: remove outdated `Item::as_searchable` implementation

cameron1024 committed:
fix: clippy

Be-ing commented:
This is a great feature, thank you @ddoemonn and @cameron1024. On the one hand it is nice that commits' diffs are filtered to just that file's history, but it would be great to have an easy way to switch between viewing just that file's history and the diff for the entire commit. Perhaps the commit hash could be turned into a link that shows the entire commit? Or a dedicated button for this?
